python/cudf/cudf/core/indexing_utils.py

```diff
@@ -241,3 +267,356 @@ def parse_row_iloc_indexer(key: Any, n: int) -> IndexingSpec:
+    This sorts the gather map so that the result comes back in the
+    order the needles were specified (and are found in the haystack).
+    """
+    # Pre-process to match dtypes
+    needle_kind = needles.dtype.kind
+    haystack_kind = haystack.dtype.kind
+    if haystack_kind == "O":
+        try:
+            needles = needles.astype(haystack.dtype)
+        except ValueError:
+            # Pandas raise KeyError here
+            raise KeyError("Dtype mismatch in label lookup")
+    elif needle_kind == haystack_kind or {
+        haystack_kind,
+        needle_kind,
+    }.issubset({"i", "u", "f"}):
+        needles = needles.astype(haystack.dtype)
+    elif needles.dtype != haystack.dtype:
+        # Pandas raise KeyError here
+        raise KeyError("Dtype mismatch in label lookup")
+    # Can't always do an inner join because then we can't check if we
+    # had missing keys (can't check the length because the entries in
+    # the needle might appear multiple times in the haystack).
+    lgather, rgather = libcudf.join.join([needles], [haystack], how="left")
+    (left_order,) = libcudf.copying.gather(
+        [cudf.core.column.arange(len(needles), dtype=size_type_dtype)],
+        lgather,
+        nullify=False,
+    )
+    (right_order,) = libcudf.copying.gather(
+        [cudf.core.column.arange(len(haystack), dtype=size_type_dtype)],
+        rgather,
+        nullify=True,
+    )
+    if right_order.null_count > 0:
+        raise KeyError("Not all keys in index")
+    (rgather,) = libcudf.sort.sort_by_key(
+        [rgather],
+        [left_order, right_order],
+        [True, True],
+        ["last", "last"],
+        stable=True,
+    )
+    return GatherMap.from_column_unchecked(
+        rgather, len(haystack), nullify=False
+    )
+
+
+def find_label_range_or_mask(
+    key: slice, index: cudf.BaseIndex
+) -> EmptyIndexer | MapIndexer | MaskIndexer | SliceIndexer:
+    """
+    Convert a slice of labels into a slice of positions
+
+    Parameters
+    ----------
+    key
+        Slice to convert
+    index
+        Index to look up in
+
+    Returns
+    -------
+    IndexingSpec
+        Structured data for indexing (but never a :class:`ScalarIndexer`)
+
+    Raises
+    ------
+    KeyError
+        If the index is unsorted and not a DatetimeIndex
+
+    Notes
+    -----
+    Until Pandas 2, looking up slices in an unsorted DatetimeIndex
+    constructs a mask by checking which dates fall in the range.
+
+    From Pandas 2, slice lookup in DatetimeIndexes will behave
+    identically to other index types and fail with a KeyError for
+    an unsorted index if either of the slice endpoints are not unique
+    in the index or are not in the index at all.
+    """
+    if (
+        not (key.start is None and key.stop is None)
+        and isinstance(index, cudf.core.index.DatetimeIndex)
+        and not index.is_monotonic_increasing
+    ):
+        # TODO: datetime index must only be handled specially until pandas 2
```

Comment:
In pandas 2, slicing unsorted datetime indexes works the same as all other indexes (although the warnings were not applied consistently, see pandas-dev/pandas#53983). So should we warning here as well? (Requires extra checking to only warn when the start or stop are not in the index).